File: sources/c/10_pi.c
// task 10 pi — expected output: 44889
// build: gcc -O2 -pthread -o prog 10_pi.c    run: ./prog
// alternates: clang -O2 -pthread -o prog 10_pi.c | cl /O2 /Fe:prog 10_pi.c | tcc -o prog 10_pi.c

#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>

/* Gibbons' unbounded spigot runs on arbitrary-precision integers, which C has no
   library for, so the state is kept in hand-written big integers: sign-magnitude,
   little-endian 64-bit limbs, base 1e9. Only the operations the spigot needs are
   here: add, subtract, multiply by a small int (limb * small + carry fits in 64
   bits for small < 2^32), and a quotient, which is always one decimal digit and so
   comes out of repeated subtraction. No 128-bit arithmetic is used, so tcc and
   MSVC build it too. The digits themselves are never printed, only their sum. */

#define BASE 1000000000ULL

typedef struct {
    uint64_t *limb;   /* little-endian, base 1e9 */
    size_t n;         /* limb count, no leading zero limbs */
    size_t cap;
    int neg;
} Big;

static void big_init(Big *x) {
    x->n = 0;
    x->cap = 8;
    x->neg = 0;
    x->limb = (uint64_t *)malloc(x->cap * sizeof(uint64_t));
}

static void big_free(Big *x) {
    free(x->limb);
    x->limb = NULL;
}

static void big_reserve(Big *x, size_t need) {
    if (need <= x->cap) {
        return;
    }
    while (x->cap < need) {
        x->cap *= 2;
    }
    x->limb = (uint64_t *)realloc(x->limb, x->cap * sizeof(uint64_t));
}

static void big_trim(Big *x) {
    while (x->n > 0 && x->limb[x->n - 1] == 0) {
        x->n--;
    }
    if (x->n == 0) {
        x->neg = 0;
    }
}

static void big_set(Big *x, uint64_t v) {
    x->n = 0;
    x->neg = 0;
    while (v > 0) {
        big_reserve(x, x->n + 1);
        x->limb[x->n++] = v % BASE;
        v /= BASE;
    }
}

static void big_copy(Big *dst, const Big *src) {
    big_reserve(dst, src->n);
    if (src->n > 0) {
        memcpy(dst->limb, src->limb, src->n * sizeof(uint64_t));
    }
    dst->n = src->n;
    dst->neg = src->neg;
}

static int big_cmp_mag(const Big *a, const Big *b) {
    if (a->n != b->n) {
        return a->n < b->n ? -1 : 1;
    }
    for (size_t i = a->n; i-- > 0; ) {
        if (a->limb[i] != b->limb[i]) {
            return a->limb[i] < b->limb[i] ? -1 : 1;
        }
    }
    return 0;
}

static int big_cmp(const Big *a, const Big *b) {
    if (a->neg != b->neg) {
        return a->neg ? -1 : 1;
    }
    int c = big_cmp_mag(a, b);
    return a->neg ? -c : c;
}

static void big_add_mag(Big *r, const Big *a, const Big *b) {
    size_t n = a->n > b->n ? a->n : b->n;
    big_reserve(r, n + 1);
    uint64_t carry = 0;
    for (size_t i = 0; i < n; i++) {
        uint64_t s = carry;
        if (i < a->n) {
            s += a->limb[i];
        }
        if (i < b->n) {
            s += b->limb[i];
        }
        if (s >= BASE) {
            s -= BASE;
            carry = 1;
        } else {
            carry = 0;
        }
        r->limb[i] = s;
    }
    r->limb[n] = carry;
    r->n = n + (carry ? 1 : 0);
    r->neg = 0;
}

static void big_sub_mag(Big *r, const Big *a, const Big *b) {   /* requires a >= b >= 0 */
    big_reserve(r, a->n);
    uint64_t borrow = 0;
    for (size_t i = 0; i < a->n; i++) {
        uint64_t bi = (i < b->n ? b->limb[i] : 0) + borrow;
        if (a->limb[i] >= bi) {
            r->limb[i] = a->limb[i] - bi;
            borrow = 0;
        } else {
            r->limb[i] = a->limb[i] + BASE - bi;
            borrow = 1;
        }
    }
    r->n = a->n;
    r->neg = 0;
    big_trim(r);
}

static void big_add(Big *r, const Big *a, const Big *b) {
    int an = a->neg, bn = b->neg;
    if (an == bn) {
        big_add_mag(r, a, b);
        r->neg = an;
    } else if (big_cmp_mag(a, b) >= 0) {
        big_sub_mag(r, a, b);
        r->neg = an;
    } else {
        big_sub_mag(r, b, a);
        r->neg = bn;
    }
    big_trim(r);
}

static void big_sub(Big *r, const Big *a, const Big *b) {       /* r = a - b */
    int an = a->neg, bn = b->neg;
    if (an != bn) {
        big_add_mag(r, a, b);
        r->neg = an;
    } else if (big_cmp_mag(a, b) >= 0) {
        big_sub_mag(r, a, b);
        r->neg = an;
    } else {
        big_sub_mag(r, b, a);
        r->neg = !an;
    }
    big_trim(r);
}

static void big_mul_small(Big *r, const Big *a, uint64_t m) {
    if (m == 0 || a->n == 0) {
        r->n = 0;
        r->neg = 0;
        return;
    }
    big_reserve(r, a->n + 2);
    uint64_t carry = 0;
    for (size_t i = 0; i < a->n; i++) {
        uint64_t p = a->limb[i] * m + carry;
        r->limb[i] = p % BASE;
        carry = p / BASE;
    }
    size_t n = a->n;
    while (carry > 0) {
        r->limb[n++] = carry % BASE;
        carry /= BASE;
    }
    r->n = n;
    r->neg = a->neg;
    big_trim(r);
}

/* floor(a / b) for a >= 0 and b > 0. The spigot only ever asks for a quotient of
   one decimal digit, so counting how many times b fits into a is enough. */
static uint64_t big_quot(const Big *a, const Big *b, Big *work) {
    uint64_t q = 0;
    if (a->neg || b->neg || b->n == 0) {
        return 0;
    }
    big_copy(work, b);
    while (big_cmp(a, work) >= 0) {
        q++;
        big_add_mag(work, work, b);
    }
    return q;
}

int main(void) {
    Big q, r, t, u, v, w;
    big_init(&q);
    big_init(&r);
    big_init(&t);
    big_init(&u);
    big_init(&v);
    big_init(&w);

    big_set(&q, 1);
    big_set(&r, 0);
    big_set(&t, 1);

    uint64_t k = 1, l = 3, n = 3;
    uint64_t sum = 0;

    for (int64_t produced = 0; produced < 10000; ) {
        big_mul_small(&u, &q, 4);
        big_add(&u, &u, &r);             /* u = 4q + r */
        big_mul_small(&v, &t, n + 1);    /* v = (n + 1)t */

        if (big_cmp(&u, &v) < 0) {
            /* the digit n is settled */
            sum += n;
            produced++;

            big_mul_small(&u, &q, 3);
            big_add(&u, &u, &r);
            big_mul_small(&u, &u, 10);           /* u = 10(3q + r) */
            uint64_t next = big_quot(&u, &t, &w) - 10 * n;

            big_mul_small(&v, &t, n);            /* v = n t */
            big_sub(&v, &r, &v);                 /* v = r - n t */
            big_mul_small(&r, &v, 10);           /* r = 10(r - n t) */
            big_mul_small(&q, &q, 10);           /* q = 10q, t is unchanged */

            n = next;
        } else {
            /* not settled yet: widen the state by one more term */
            big_mul_small(&u, &q, 7 * k + 2);
            big_mul_small(&v, &r, l);
            big_add(&u, &u, &v);                 /* u = q(7k + 2) + r l */
            big_mul_small(&v, &t, l);            /* v = t l */
            uint64_t next = big_quot(&u, &v, &w);

            big_mul_small(&u, &q, 2);
            big_add(&u, &u, &r);
            big_mul_small(&u, &u, l);            /* u = (2q + r) l */
            big_copy(&r, &u);
            big_mul_small(&q, &q, k);
            big_mul_small(&t, &t, l);

            k++;
            l += 2;
            n = next;
        }
    }

    printf("%llu\n", (unsigned long long)sum);

    big_free(&q);
    big_free(&r);
    big_free(&t);
    big_free(&u);
    big_free(&v);
    big_free(&w);
    return 0;
}

File: sources/rust/10_pi.rs
// task 10 pi — expected output: 44889
// build: rustc -O -o prog 10_pi.rs    run: ./prog
//
// Gibbons' unbounded spigot, summing the first 10000 digits it emits (leading 3 included)
// and never printing them. Rust's standard library has no big integers, so the state
// (q, r, t, k, n, l) is carried in base-1e9 limbs: multiply-by-small-int and add/sub are
// hand-written below, and the two divisions (by the big t and by t*l) take their small
// quotient by binary search on q*den <= num, which needs nothing but that same
// multiply-by-small-int and a comparison.

use std::cmp::Ordering;

const BASE: u64 = 1_000_000_000;
const DIGITS: usize = 10_000;
// The spigot only ever divides by the big t (or t*l), and always for a quotient that is a
// digit or a digit times ten: floor(10*(3q+r)/t) = next_digit + 10*previous_digit <= 99,
// and the next digit itself is 0..9. Both bounds are exclusive and have room to spare.
const BOUND_SCALE: u64 = 256;
const BOUND_DIGIT: u64 = 32;

struct Big {
    neg: bool,
    limbs: Vec<u64>, // little-endian, base 1e9, no leading zeros, empty means zero
}

impl Big {
    fn norm(neg: bool, mut limbs: Vec<u64>) -> Big {
        while let Some(&0) = limbs.last() {
            limbs.pop();
        }
        Big { neg: neg && !limbs.is_empty(), limbs }
    }

    fn zero() -> Big {
        Big { neg: false, limbs: Vec::new() }
    }

    fn from_u64(mut n: u64) -> Big {
        let mut limbs = Vec::new();
        while n > 0 {
            limbs.push(n % BASE);
            n /= BASE;
        }
        Big { neg: false, limbs }
    }

    fn is_zero(&self) -> bool {
        self.limbs.is_empty()
    }

    fn cmp(&self, other: &Big) -> Ordering {
        if self.neg != other.neg {
            return if self.neg { Ordering::Less } else { Ordering::Greater };
        }
        let c = mag_cmp(&self.limbs, &other.limbs);
        if self.neg { c.reverse() } else { c }
    }

    fn add(&self, other: &Big) -> Big {
        if self.neg == other.neg {
            Big::norm(self.neg, mag_add(&self.limbs, &other.limbs))
        } else {
            match mag_cmp(&self.limbs, &other.limbs) {
                Ordering::Greater => Big::norm(self.neg, mag_sub(&self.limbs, &other.limbs)),
                Ordering::Less => Big::norm(other.neg, mag_sub(&other.limbs, &self.limbs)),
                Ordering::Equal => Big::zero(),
            }
        }
    }

    fn sub(&self, other: &Big) -> Big {
        if self.neg != other.neg {
            Big::norm(self.neg, mag_add(&self.limbs, &other.limbs))
        } else {
            match mag_cmp(&self.limbs, &other.limbs) {
                Ordering::Greater => Big::norm(self.neg, mag_sub(&self.limbs, &other.limbs)),
                Ordering::Less => Big::norm(!self.neg, mag_sub(&other.limbs, &self.limbs)),
                Ordering::Equal => Big::zero(),
            }
        }
    }

    fn mul_small(&self, m: u64) -> Big {
        if m == 0 || self.is_zero() {
            return Big::zero();
        }
        let mut out = Vec::with_capacity(self.limbs.len() + 2);
        let mut carry = 0u64;
        for &limb in &self.limbs {
            let p = limb * m + carry;
            out.push(p % BASE);
            carry = p / BASE;
        }
        while carry > 0 {
            out.push(carry % BASE);
            carry /= BASE;
        }
        Big::norm(self.neg, out)
    }

    /// floor(self / den), for den > 0 and |result| < bound.
    fn div_floor(&self, den: &Big, bound: u64) -> i64 {
        if self.neg {
            -(smallest_ge(&self.limbs, &den.limbs, bound) as i64)
        } else {
            largest_le(&self.limbs, &den.limbs, bound) as i64
        }
    }
}

fn mag_cmp(a: &[u64], b: &[u64]) -> Ordering {
    if a.len() != b.len() {
        return a.len().cmp(&b.len());
    }
    for i in (0..a.len()).rev() {
        if a[i] != b[i] {
            return a[i].cmp(&b[i]);
        }
    }
    Ordering::Equal
}

fn mag_add(a: &[u64], b: &[u64]) -> Vec<u64> {
    let (long, short) = if a.len() >= b.len() { (a, b) } else { (b, a) };
    let mut out = Vec::with_capacity(long.len() + 1);
    let mut carry = 0u64;
    for i in 0..long.len() {
        let s = long[i] + short.get(i).copied().unwrap_or(0) + carry;
        out.push(s % BASE);
        carry = s / BASE;
    }
    if carry > 0 {
        out.push(carry);
    }
    out
}

/// a - b, requires a >= b.
fn mag_sub(a: &[u64], b: &[u64]) -> Vec<u64> {
    let mut out = Vec::with_capacity(a.len());
    let mut borrow = 0u64;
    for i in 0..a.len() {
        let bi = b.get(i).copied().unwrap_or(0);
        if a[i] >= bi + borrow {
            out.push(a[i] - bi - borrow);
            borrow = 0;
        } else {
            out.push(a[i] + BASE - bi - borrow);
            borrow = 1;
        }
    }
    while let Some(&0) = out.last() {
        out.pop();
    }
    out
}

/// compare q * d with the magnitude a, q being a small multiplier.
fn cmp_mul_mag(d: &[u64], q: u64, a: &[u64]) -> Ordering {
    let mut carry = 0u64;
    let mut i = 0usize;
    while i < d.len() || carry > 0 {
        let di = if i < d.len() { d[i] } else { 0 };
        let p = di * q + carry;
        let limb = p % BASE;
        carry = p / BASE;
        let ai = if i < a.len() { a[i] } else { 0 };
        if limb != ai {
            return limb.cmp(&ai);
        }
        i += 1;
    }
    while i < a.len() {
        if a[i] != 0 {
            return Ordering::Less;
        }
        i += 1;
    }
    Ordering::Equal
}

/// largest q < bound with q * d <= a.
fn largest_le(a: &[u64], d: &[u64], bound: u64) -> u64 {
    let mut lo = 0u64;
    let mut hi = bound;
    while lo + 1 < hi {
        let mid = lo + (hi - lo) / 2;
        if cmp_mul_mag(d, mid, a) == Ordering::Greater {
            hi = mid;
        } else {
            lo = mid;
        }
    }
    lo
}

/// smallest q < bound with q * d >= a.
fn smallest_ge(a: &[u64], d: &[u64], bound: u64) -> u64 {
    let mut lo = 0u64;
    let mut hi = bound;
    while lo < hi {
        let mid = lo + (hi - lo) / 2;
        if cmp_mul_mag(d, mid, a) == Ordering::Less {
            lo = mid + 1;
        } else {
            hi = mid;
        }
    }
    lo
}

fn main() {
    let mut q = Big::from_u64(1);
    let mut r = Big::zero();
    let mut t = Big::from_u64(1);
    let mut k: u64 = 1;
    let mut n: i64 = 3;
    let mut l: u64 = 3;

    let mut sum: i64 = 0;
    let mut emitted: usize = 0;

    while emitted < DIGITS {
        // 4*q + r - t < n*t ?
        let lhs = q.mul_small(4).add(&r).sub(&t);
        let rhs = t.mul_small(n as u64);

        if lhs.cmp(&rhs) == Ordering::Less {
            // the digit is safe: emit n and advance
            sum += n;
            emitted += 1;
            let next_n = q.mul_small(3).add(&r).mul_small(10).div_floor(&t, BOUND_SCALE) - 10 * n;
            q = q.mul_small(10);
            r = r.sub(&t.mul_small(n as u64)).mul_small(10);
            n = next_n;
        } else {
            let next_n = q
                .mul_small(7 * k + 2)
                .add(&r.mul_small(l))
                .div_floor(&t.mul_small(l), BOUND_DIGIT);
            let next_r = q.mul_small(2).add(&r).mul_small(l);
            let next_q = q.mul_small(k);
            let next_t = t.mul_small(l);
            q = next_q;
            r = next_r;
            t = next_t;
            k += 1;
            n = next_n;
            l += 2;
        }
    }

    println!("{}", sum);
}
